### ISSUE TITLE
Error when testing equality of AdditiveSharingTensors (without a crypto_provider?)

### ISSUE THREAD

**Issue by kennysong** [OpenMined/PySyft#3540]
When I try to test equality of two AdditiveSharingTensors (`x == y`), it fails with a `ObjectNotFoundError` error. 

Here's a basic code sample to reproduce:

```python
import syft as sy
import torch as th
hook = sy.TorchHook(th)

bob = sy.VirtualWorker(hook, id="bob")
alice = sy.VirtualWorker(hook, id="alice")

x = th.tensor([1,2,3,4])
y = th.tensor([2,-1,1,0])
x = x.share(bob, alice)
y = y.share(bob, alice)

z = x == y
z.get()
```

This fails on the `==` line with: 

```
ObjectNotFoundError: Object "7070927336" not found on worker!!! You just tried to interact with an object ID:7070927336 on <VirtualWorker id:me #objects:0> which does not exist!!! Use .send() and .get() on all your tensors to make sure they're on the same machines. If you think this tensor does exist, check the ._objects dictionary on the worker and see for yourself!!! The most common reason this error happens is because someone calls .get() on the object's pointer without realizing it (which deletes the remote object and sends it to the pointer). Check your code to make sure you haven't already called .get() on this pointer!!!
```

However, this seems to work if I add a separate worker as the `crypto_provider`. For example, if you replace the two lines with `.share()` above with:

```python
secure_worker = sy.VirtualWorker(hook, id="secure_worker")
x = x.share(bob, alice, crypto_provider=secure_worker)
y = y.share(bob, alice, crypto_provider=secure_worker)
```

^This works. Is this a bug? Or intended behavior (where is it documented)?

**Comment by kennysong** [comment]
Same problem for multiplication!

```python
import syft as sy
import torch as th
hook = sy.TorchHook(th)

bob = sy.VirtualWorker(hook, id="bob")
alice = sy.VirtualWorker(hook, id="alice")

x = th.tensor([1,2,3,4])
y = th.tensor([2,-1,1,0])
x = x.share(bob, alice)
y = y.share(bob, alice)

z = x * y  # <-- ObjectNotFoundError
z.get()
```

Only addition (`x + y`) works without specifying a `crypto_provider`.

**Comment by sukhadj** [comment]
> Same problem for multiplication!
> 
> ```python
> import syft as sy
> import torch as th
> hook = sy.TorchHook(th)
> 
> bob = sy.VirtualWorker(hook, id="bob")
> alice = sy.VirtualWorker(hook, id="alice")
> 
> x = th.tensor([1,2,3,4])
> y = th.tensor([2,-1,1,0])
> x = x.share(bob, alice)
> y = y.share(bob, alice)
> 
> z = x * y  # <-- ObjectNotFoundError
> z.get()
> ```
> 
> Only addition (`x + y`) works without specifying a `crypto_provider`.

Hey, so AST uses secure multiparty communication (SMPC) to perform the operations like multiplication, division. It uses `crypto provider` to generate the triplets which are necessary to perform this operations. So crypto provider is kind of necessary 🙂

**Comment by karlhigley** [comment]
It seems like `share()` is overly permissive then. Until recently (when @Prtfw added an exception), `share()` could be called with no arguments. Something similar should probably happen here: a required argument that's missing should immediately raise an exception rather than allowing incorrect behavior.

**Comment by kennysong** [comment]
Thanks @sukhadj, I figured it was something like that. A few beginner questions:

1. What is AST?
2. Why does `x + y` work without a crypto provider and not `x * y`?
3. Can I read more about this in the documentation somewhere?
4. If `crypto_provider` is not always required for `.share()`, as an alternative to @karlhigley's suggestion, can the error message be more specific? e.g. `crypto_provider required for * operation on AdditiveSharingTensors` rather than an `ObjectNotFoundError`.

**Comment by AlanAboudib** [comment]
I also had this equality issue, even with a crypto provider, only when `requires_grad = False`

**Comment by sukhadj** [comment]
> Thanks @sukhadj, I figured it was something like that. A few beginner questions:
> 
>     1. What is AST?
AdditiveSharingTensor
>     2. Why does `x + y` work without a crypto provider and not `x * y`?
So addition is just basically adding alice's share of x with alice's share of y and then module it. You can see implementation of addition [here](https://github.com/OpenMined/PySyft/blob/6b55ba01e411537e5c43f44235d24c33bdbaadf0/syft/frameworks/torch/tensors/interpreters/additive_shared.py#L413).
>     3. Can I read more about this in the documentation somewhere?
PySyft currently don't have extensive documentation right now. But you can find more about SPDZ and triplets [here](https://mortendahl.github.io/2017/09/03/the-spdz-protocol-part1/) (Not an expert of the subject so maybe others can suggest better resources maybe :smile: )
>     4. If `crypto_provider` is not always required for `.share()`, as an alternative to @karlhigley's suggestion, can the error message be more specific? e.g. `crypto_provider required for * operation on AdditiveSharingTensors` rather than an `ObjectNotFoundError`.
That's correct we should definitely raise a warning during the sharing and exception when multiplying rather than object not found error.

**Comment by sukhadj** [comment]
> I also had this equality issue, even with a crypto provider, only when `requires_grad = False`
Can you try again with pulling the master?
It works now when provided with crypto provider. Maybe changed in @LaRiffle's  FSS PR.

**Comment by AlanAboudib** [comment]
Does `==` work if you set `requires_grad = False` in the `share` method?

**Comment by LaRiffle** [comment]
Will be solved by #3587
